sys/nhdp: Add metric infrastructure and DAT metric computation

fnack opened the pull request:
This PR extends NHDP's functionality with the possibility to calculate and exchange link metric values for the detected links.

The PR depends on ~~#2500~~ and ~~#2576~~.

emmanuelsearch commented:
@fnack I guess the other PRs on which this PR were merged. So how about updating this one?

emmanuelsearch commented:
@fnack mention in the PR name that it adds support for DAT (Directional Airtime Metric [1]).
-- if I'm not mistaken ;)

[1] https://tools.ietf.org/html/draft-ietf-manet-olsrv2-dat-metric-05

fnack commented:
Rebased. I'll update the PR over the weekend and provide an example application.

OlegHahm commented:
Have you updated? If yes, please do not squash immediately and leave a notice the next time. If no, any new time estimate?

fnack commented:
Yes, I updated. Sorry. You could start to review, only thing that should be missing is documentation for the macros. I'll add them when I'm back in Berlin today and rebase the application  I provided for the other PRs.

fnack commented:
Ready for review. I added documentation for the newly introduced macros and squashed right away because no code changes were included.

An example application including the changes of this PR can be found [here](https://github.com/RIOT-OS/RIOT/compare/master...fnack:nhdp_app_opt). Change the `NHDP_METRIC` macro to switch the used metric.

OlegHahm commented:
Great. Will review ASAP.

OlegHahm commented on sys/net/routing/nhdp/iib_table.c:
```diff
@@ -468,6 +609,18 @@ static void reset_link_set_entry(iib_link_set_entry_t *ls_entry, timex_t *now, u
+#if (NHDP_METRIC == NHDP_LMT_DAT)
+    memset(ls_entry->dat_received, 0, NHDP_Q_MEM_LENGTH);
+    memset(ls_entry->dat_total, 0, NHDP_Q_MEM_LENGTH);
+    ls_entry->dat_time.microseconds = 0;
```
Sorry, if I asked this already, but is microsecond precision required?

fnack replied:
No, but I used timex values to be able to use functionality like `vtimer_now` or `timex_add`. Of course there is great potential to save a lot of memory but that would be a complex optimization step.

OlegHahm replied:
But is this really this complex here? If you say yes, we can just leave it for future improvement PRs.

OlegHahm commented:
Maybe you could put a link to the current Internet drafts or RFCs somewhere in the documentation of the header.

OlegHahm commented on sys/net/routing/nhdp/nhdp_metric.h:
```diff
@@ -0,0 +1,88 @@
+#define NHDP_LMT_DAT                (165)
+
+/** @brief Used metric (change to switch to another metric type) */
+#define NHDP_METRIC                 NHDP_LMT_HOP_COUNT
```
Parens.

OlegHahm commented on sys/net/routing/nhdp/iib_table.c:
```diff
@@ -216,6 +234,138 @@ void iib_propagate_nb_entry_change(nib_entry_t *old_entry, nib_entry_t *new_entr
+            }
+            else {
+                /* Don't add values to the queue for duplicate packets */
+                if (seq_no != ls_entry->last_seq_no) {
```
`else if () {`?

OlegHahm commented on sys/net/routing/nhdp/iib_table.c:
```diff
@@ -216,6 +234,138 @@ void iib_propagate_nb_entry_change(nib_entry_t *old_entry, nib_entry_t *new_entr
+                if (seq_no != ls_entry->last_seq_no) {
+                    uint16_t seq_diff;
+                    if (seq_no < ls_entry->last_seq_no) {
+                        seq_diff = (uint16_t) ((((uint32_t) seq_no) + 0xFFFF) - ls_entry->last_seq_no);
```
Wouldn`t be
```C
seq_diff = ls_entry->last_seq - seq_no;
```
much easier to read?

fnack replied:
This would result in values that are not treated correctly by the restart detection.

The DAT draft states:
>   diff_seqno(new, old)  - an operation which returns the positive
      distance between two elements of the circular sequence number
      space defined in section 5.1 of [RFC5444].  Its value is either
      (new - old) if this result is positive, or else its value is (new
      - old + 65536).

Of course I could adapt the restart detection, but this would result in adding complexity somewhere else.

OlegHahm replied:
Ok, I think I misinterpreted the line and if the draft is that explicit, then we should probably leave it.

OlegHahm commented on sys/net/routing/nhdp/iib_table.c:
```diff
@@ -216,6 +234,138 @@ void iib_propagate_nb_entry_change(nib_entry_t *old_entry, nib_entry_t *new_entr
+
+void iib_process_metric_refresh(void)
+{
+    switch (NHDP_METRIC) {
```
Switch case on a define is a bit pointless, isn't it? (I mean this can be done at build time or am I missing something?)

fnack replied:
That's true. It looked a little easier to understand in my opinion, but I'll change it to reduce code size/complexity.

OlegHahm commented on sys/net/routing/nhdp/iib_table.c:
```diff
@@ -665,3 +827,111 @@ static inline timex_t get_max_timex(timex_t time_one, timex_t time_two)
+{
+    uint16_t sum = 0;
+
+    for (int i=0; i < NHDP_Q_MEM_LENGTH; i++) {
```
spaces around `=`

OlegHahm commented on sys/net/routing/nhdp/iib_table.h:
```diff
@@ -62,6 +74,8 @@ typedef struct iib_two_hop_set_entry_t {
     nhdp_addr_t *th_nb_addr;                    /**< Address of symmetric 2-hop neighbor */
     timex_t exp_time;                           /**< Time at which entry expires */
+    uint32_t metric_in;                         /**< Metric value for incoming link */
+    uint32_t metric_out;                        /**< Metric value for outgoing link */
```
Too lazy to look it up, but are these values really specified as being 32 bit?

fnack replied:
The RFC of OLSRv2 states
   >Link metrics are reported in messages using a compressed
   representation that occupies 12 bits, consisting of a 4-bit field and
   an 8-bit field.  The compressed representation represents positive
   integer values with a minimum value of 1 and a maximum value that is
   slightly smaller than the maximum 24-bit value.  Only those values
   that have exact representation in the compressed form are used.
   Route metrics are the summation of no more than 256 link metric
   values and can therefore be represented using no more than 32 bits.
   
   >Link and route metrics used in the Information Bases defined in this
   specification refer to the uncompressed values, and arithmetic
   involving them does likewise and assumes full precision in the
   result.  (How an implementation records the values is not part of
   this specification, as long as it behaves as if recording
   uncompressed values.  An implementation can, for example, use 32-bit
   values for all link and route metrics.)

An uncompressed metric value therefore needs up to 24 bits.

OlegHahm replied:
Hm, okay, too bad.

OlegHahm commented on sys/net/routing/nhdp/nhdp_reader.c:
```diff
@@ -186,12 +222,25 @@ _nhdp_blocktlv_address_cb(struct rfc5444_reader_tlvblock_context *cont)
 
+        if (lt_metric_val == NHDP_METRIC_UNKNOWN) {
+            /* Determine our outgoing link metric value to the originator interface */
+            if (_nhdp_addr_tlvs[RFC5444_ADDRTLV_LINK_METRIC].tlv) {
```
aggregate these two ifs

OlegHahm commented:
@fnack, any time estimation to address the latest comments? I think this is almost ready for merging.

fnack commented:
Sorry, got a little lazy with less spare time. I'll update today, when I'm back from work.

OlegHahm commented:
No need to apologize. ;)

fnack commented:
I rebased and addressed your comments.

OlegHahm commented:
Untested ACK. Will test now, but you can already squash. Will open an issue for the timex thing.

OlegHahm commented:
Tested ACK. Squash and merge as soon as Travis is okay.

fnack commented:
Squashed, waiting for Travis.

fnack commented:
Travis is happy, go...
